lib/internal/wasi.js

```diff
@@ -0,0 +1,1301 @@
+  [WASI_SIGQUIT]: 'SIGQUIT',
+  [WASI_SIGILL]: 'SIGILL',
+  [WASI_SIGTRAP]: 'SIGTRAP',
+  [WASI_SIGABRT]: 'SIGABRT',
+  // [WASI_SIGIOT]: 'SIGIOT',
+  [WASI_SIGBUS]: 'SIGBUS',
+  [WASI_SIGFPE]: 'SIGFPE',
+  [WASI_SIGKILL]: 'SIGKILL',
+  [WASI_SIGUSR1]: 'SIGUSR1',
+  [WASI_SIGSEGV]: 'SIGSEGV',
+  [WASI_SIGUSR2]: 'SIGUSR2',
+  [WASI_SIGPIPE]: 'SIGPIPE',
+  [WASI_SIGALRM]: 'SIGALRM',
+  [WASI_SIGTERM]: 'SIGTERM',
+  [WASI_SIGCHLD]: 'SIGCHLD',
+  [WASI_SIGCONT]: 'SIGCONT',
+  [WASI_SIGSTOP]: 'SIGSTOP',
+  [WASI_SIGTSTP]: 'SIGTSTP',
+  [WASI_SIGTTIN]: 'SIGTTIN',
+  [WASI_SIGTTOU]: 'SIGTTOU',
+  [WASI_SIGURG]: 'SIGURG',
+  [WASI_SIGXCPU]: 'SIGXCPU',
+  [WASI_SIGXFSZ]: 'SIGXFSZ',
+  [WASI_SIGVTALRM]: 'SIGVTALRM',
+  // [WASI_SIGPROF]: 'SIGPROF',
+  // [WASI_SIGWINCH]: 'SIGWINCH',
+  // [WASI_SIGIO]: 'SIGIO',
+  // [WASI_SIGINFO]: 'SIGINFO',
+  // [WASI_SIGSYS]: 'SIGSYS',
+};
+
+const now = (clockId) => {
+  switch (clockId) {
+    case WASI_CLOCK_MONOTONIC:
+    case WASI_CLOCK_REALTIME:
+      return process.hrtime.bigint();
+    case WASI_CLOCK_PROCESS_CPUTIME_ID:
+    case WASI_CLOCK_THREAD_CPUTIME_ID:
+      return process.hrtime.bigint(CPUTIME_START);
+    default:
+      return null;
+  }
+};
+
+
+const msToNs = (ms) => {
+  const msInt = Math.trunc(ms);
+  const decimal = BigInt(Math.round((ms - msInt) * 1000));
+  const ns = BigInt(msInt) * 1000n;
+  return ns + decimal;
+};
+
+const wrap = (f) => (...args) => {
+  try {
+    return f(...args);
+  } catch (e) {
+    if (typeof e === 'number') {
+      return e;
+    }
+    if (e.errno) {
+      return ERROR_MAP[e.code] || WASI_EINVAL;
+    }
+    throw e;
+  }
+};
+
+const translateFileAttributes = (fd, stats) => {
+  switch (true) {
+    case stats.isBlockDevice():
+      return {
+        filetype: WASI_FILETYPE_BLOCK_DEVICE,
+        rightsBase: RIGHTS_BLOCK_DEVICE_BASE,
+        rightsInheriting: RIGHTS_BLOCK_DEVICE_INHERITING,
+      };
+    case stats.isCharacterDevice(): {
+      const filetype = WASI_FILETYPE_CHARACTER_DEVICE;
+      if (fd !== undefined && isTTY(fd)) {
+        return {
+          filetype,
+          rightsBase: RIGHTS_TTY_BASE,
+          rightsInheriting: RIGHTS_TTY_INHERITING,
+        };
+      }
+      return {
+        filetype,
+        rightsBase: RIGHTS_CHARACTER_DEVICE_BASE,
+        rightsInheriting: RIGHTS_CHARACTER_DEVICE_INHERITING,
+      };
+    }
+    case stats.isDirectory():
+      return {
+        filetype: WASI_FILETYPE_DIRECTORY,
+        rightsBase: RIGHTS_DIRECTORY_BASE,
+        rightsInheriting: RIGHTS_DIRECTORY_INHERITING,
+      };
+    case stats.isFIFO():
+      return {
+        filetype: WASI_FILETYPE_SOCKET_STREAM,
+        rightsBase: RIGHTS_SOCKET_BASE,
+        rightsInheriting: RIGHTS_SOCKET_INHERITING,
+      };
+    case stats.isFile():
+      return {
+        filetype: WASI_FILETYPE_REGULAR_FILE,
+        rightsBase: RIGHTS_REGULAR_FILE_BASE,
+        rightsInheriting: RIGHTS_REGULAR_FILE_INHERITING,
```

Comment:
These do not represent the actual access rights, right? Is that okay? (I think we would need to support `fcntl()` better here, so assuming that this is just a temporary solution, I think a `// TODO` comment would be appropriate here as well)